File: packages/sdk-js/src/utils/http.ts
import axios, { AxiosInstance, AxiosError, AxiosRequestConfig } from 'axios';
import type { ApiError, LnkClientConfig } from '../types';

export class HttpClient {
  private client: AxiosInstance;
  private config: LnkClientConfig;
  private accessToken?: string;
  private refreshToken?: string;
  private tokenExpiresAt?: number;

  constructor(config: LnkClientConfig) {
    this.config = config;
    this.accessToken = config.accessToken;

    this.client = axios.create({
      baseURL: config.baseUrl || 'https://api.lnk.day',
      timeout: config.timeout || 30000,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': '@lnk/sdk/1.0.0',
      },
    });

    this.setupInterceptors();
  }

  private setupInterceptors(): void {
    // Request interceptor - add auth headers
    this.client.interceptors.request.use(
      (config) => {
        if (this.config.apiKey) {
          config.headers['X-API-Key'] = this.config.apiKey;
        } else if (this.accessToken) {
          config.headers['Authorization'] = `Bearer ${this.accessToken}`;
        }
        return config;
      },
      (error) => Promise.reject(error)
    );

    // Response interceptor - handle errors and token refresh
    this.client.interceptors.response.use(
      (response) => response,
      async (error: AxiosError) => {
        const originalRequest = error.config as AxiosRequestConfig & { _retry?: boolean };

        // Handle 401 with token refresh
        if (
          error.response?.status === 401 &&
          this.refreshToken &&
          !originalRequest._retry
        ) {
          originalRequest._retry = true;
          try {
            await this.refreshAccessToken();
            if (originalRequest.headers) {
              originalRequest.headers['Authorization'] = `Bearer ${this.accessToken}`;
            }
            return this.client(originalRequest);
          } catch (refreshError) {
            return Promise.reject(this.transformError(error));
          }
        }

        return Promise.reject(this.transformError(error));
      }
    );
  }

  private transformError(error: AxiosError): ApiError {
    if (error.response) {
      const data = error.response.data as any;
      return {
        statusCode: error.response.status,
        message: data?.message || error.message,
        error: data?.error,
        details: data?.details,
      };
    }

    if (error.request) {
      return {
        statusCode: 0,
        message: 'Network error - no response received',
        error: 'NETWORK_ERROR',
      };
    }

    return {
      statusCode: 0,
      message: error.message || 'Unknown error',
      error: 'UNKNOWN_ERROR',
    };
  }

  private async refreshAccessToken(): Promise<void> {
    if (!this.refreshToken) {
      throw new Error('No refresh token available');
    }

    const response = await axios.post(
      `${this.config.baseUrl || 'https://api.lnk.day'}/auth/refresh`,
      { refreshToken: this.refreshToken }
    );

    this.accessToken = response.data.accessToken;
    this.refreshToken = response.data.refreshToken;
    this.tokenExpiresAt = response.data.expiresAt;

    if (this.config.onTokenRefresh) {
      this.config.onTokenRefresh(this.accessToken!);
    }
  }

  setTokens(accessToken: string, refreshToken?: string, expiresAt?: number): void {
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.tokenExpiresAt = expiresAt;
  }

  clearTokens(): void {
    this.accessToken = undefined;
    this.refreshToken = undefined;
    this.tokenExpiresAt = undefined;
  }

  async get<T>(url: string, params?: Record<string, any>): Promise<T> {
    const response = await this.client.get<T>(url, { params });
    return response.data;
  }

  async post<T>(url: string, data?: any): Promise<T> {
    const response = await this.client.post<T>(url, data);
    return response.data;
  }

  async put<T>(url: string, data?: any): Promise<T> {
    const response = await this.client.put<T>(url, data);
    return response.data;
  }

  async patch<T>(url: string, data?: any): Promise<T> {
    const response = await this.client.patch<T>(url, data);
    return response.data;
  }

  async delete<T>(url: string): Promise<T> {
    const response = await this.client.delete<T>(url);
    return response.data;
  }
}

File: packages/sdk-python/lnk_sdk/http.py
"""HTTP client for lnk-sdk."""

from typing import Any, Dict, Optional, TypeVar
import httpx
from .types import ApiError

T = TypeVar("T")


class HttpClient:
    """HTTP client with authentication and error handling."""

    def __init__(
        self,
        base_url: str = "https://api.lnk.day",
        api_key: Optional[str] = None,
        access_token: Optional[str] = None,
        timeout: float = 30.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.access_token = access_token
        self.refresh_token: Optional[str] = None
        self.timeout = timeout

        self._client = httpx.Client(
            base_url=self.base_url,
            timeout=timeout,
            headers=self._get_headers(),
        )
        self._async_client: Optional[httpx.AsyncClient] = None

    def _get_headers(self) -> Dict[str, str]:
        headers = {
            "Content-Type": "application/json",
            "User-Agent": "lnk-sdk-python/1.0.0",
        }
        if self.api_key:
            headers["X-API-Key"] = self.api_key
        elif self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"
        return headers

    def set_tokens(
        self,
        access_token: str,
        refresh_token: Optional[str] = None,
    ) -> None:
        self.access_token = access_token
        self.refresh_token = refresh_token
        self._client.headers.update({"Authorization": f"Bearer {access_token}"})

    def clear_tokens(self) -> None:
        self.access_token = None
        self.refresh_token = None
        if "Authorization" in self._client.headers:
            del self._client.headers["Authorization"]

    def _handle_response(self, response: httpx.Response) -> Any:
        if response.status_code >= 400:
            try:
                data = response.json()
                raise ApiError(
                    status_code=response.status_code,
                    message=data.get("message", response.text),
                    error=data.get("error"),
                    details=data.get("details"),
                )
            except (ValueError, KeyError):
                raise ApiError(
                    status_code=response.status_code,
                    message=response.text or "Unknown error",
                )

        if response.status_code == 204:
            return None

        try:
            return response.json()
        except ValueError:
            return response.content

    def get(self, url: str, params: Optional[Dict[str, Any]] = None) -> Any:
        response = self._client.get(url, params=params)
        return self._handle_response(response)

    def post(self, url: str, data: Optional[Dict[str, Any]] = None) -> Any:
        response = self._client.post(url, json=data)
        return self._handle_response(response)

    def put(self, url: str, data: Optional[Dict[str, Any]] = None) -> Any:
        response = self._client.put(url, json=data)
        return self._handle_response(response)

    def patch(self, url: str, data: Optional[Dict[str, Any]] = None) -> Any:
        response = self._client.patch(url, json=data)
        return self._handle_response(response)

    def delete(self, url: str) -> Any:
        response = self._client.delete(url)
        return self._handle_response(response)

    # Async methods
    async def _get_async_client(self) -> httpx.AsyncClient:
        if self._async_client is None:
            self._async_client = httpx.AsyncClient(
                base_url=self.base_url,
                timeout=self.timeout,
                headers=self._get_headers(),
            )
        return self._async_client

    async def aget(self, url: str, params: Optional[Dict[str, Any]] = None) -> Any:
        client = await self._get_async_client()
        response = await client.get(url, params=params)
        return self._handle_response(response)

    async def apost(self, url: str, data: Optional[Dict[str, Any]] = None) -> Any:
        client = await self._get_async_client()
        response = await client.post(url, json=data)
        return self._handle_response(response)

    async def apatch(self, url: str, data: Optional[Dict[str, Any]] = None) -> Any:
        client = await self._get_async_client()
        response = await client.patch(url, json=data)
        return self._handle_response(response)

    async def adelete(self, url: str) -> Any:
        client = await self._get_async_client()
        response = await client.delete(url)
        return self._handle_response(response)

    def close(self) -> None:
        self._client.close()
        if self._async_client:
            # Note: async client should be closed with await
            pass

    async def aclose(self) -> None:
        self._client.close()
        if self._async_client:
            await self._async_client.aclose()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.aclose()
